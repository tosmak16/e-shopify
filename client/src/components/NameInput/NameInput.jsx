import { useState } from 'react';
import TextInput, { INPUT_TYPES } from '../TextInput/TextInput';

const NameInput = props => {
  const [name, setName] = useState({
    value: '',
    error: ''
  });

  const validateName = (value, placeholder = 'This field') => {
    let error;
    if (!value.length) {
      error = `${placeholder} can not be empty`;
    } else if (value.length < 2) {
      error = `${placeholder} should at least be two characters`;
    } else if (!/^[a-zA-Z '.-]*$/.test(value)) {
      error = `${placeholder} is invalid`;
    } else {
      error = '';
    }
    return error;
  };

  const inputChange = event => {
    const { placeholder, value } = event.target;
    const { onChange } = props;
    const error = validateName(value, placeholder);
    // set state
    setName({ error, value });
    // pass change to parent component
    return onChange({ value, error, isValid: !error.length });
  };

  const { value, error } = name;

  return (
    <div>
      <TextInput
        {...props}
        type={INPUT_TYPES.TEXT}
        isValid={!error.length}
        onChange={inputChange}
        value={value}
        errorMessage={error}
      />
    </div>
  );
};

NameInput.propTypes = {
  onChange: PropTypes.func.isRequired
};

export default NameInput;
