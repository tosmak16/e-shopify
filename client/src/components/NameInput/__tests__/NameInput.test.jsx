import NameInput from '../NameInput';
import { INPUT_TYPES } from '../../TextInput/TextInput';

describe('NameInput', () => {
	it('renders', () => {
		const tree = shallow(<NameInput id="test-input" onChange={jest.fn()} />);

		expect(tree).toMatchSnapshot();
	});

	it('should be able identify valid text', () => {
		const onChange = jest.fn();
		const tree = shallow(
			<NameInput onChange={onChange} isValid={false} value="" id="test-input" />
		);
		tree.find('#test-input').simulate('change', {
			target: {
				value: 'John',
				type: INPUT_TYPES.TEXT
			}
		});

		expect(onChange).toBeCalledWith({
			error: '',
			isValid: true,
			value: 'John'
		});
	});

	it('should validate text emptyness', () => {
		const onChange = jest.fn();
		const tree = shallow(
			<NameInput
				type={INPUT_TYPES.TEXT}
				onChange={onChange}
				isValid={false}
				value=""
				id="test-input"
			/>
		);
		tree.find('#test-input').simulate('change', {
			target: {
				value: '',
				placeholder: 'First name',
				type: INPUT_TYPES.TEXT
			}
		});

		expect(onChange).toBeCalledWith({
			error: 'First name can not be empty',
			isValid: false,
			value: ''
		});
	});

	it('should validate text less than two characters', () => {
		const onChange = jest.fn();
		const tree = shallow(
			<NameInput
				type={INPUT_TYPES.TEXT}
				onChange={onChange}
				isValid={false}
				value=""
				id="test-input"
			/>
		);
		tree.find('#test-input').simulate('change', {
			target: {
				value: 'J',
				placeholder: 'First name',
				type: INPUT_TYPES.TEXT
			}
		});

		expect(onChange).toBeCalledWith({
			error: 'First name should at least be two characters',
			isValid: false,
			value: 'J'
		});
	});

	it('should validate invalid characters in text', () => {
		const onChange = jest.fn();
		const tree = shallow(
			<NameInput
				type={INPUT_TYPES.TEXT}
				onChange={onChange}
				isValid={false}
				value=""
				id="test-input"
			/>
		);
		tree.find('#test-input').simulate('change', {
			target: {
				value: 'J@##  1',
				placeholder: 'First name',
				type: INPUT_TYPES.TEXT
			}
		});
		expect(onChange).toBeCalledWith({
			error: 'First name is invalid',
			isValid: false,
			value: 'J@##  1'
		});
	});

	it('should default to `This field` in case a placeholder is not provided', () => {
		const onChange = jest.fn();
		const tree = shallow(
			<NameInput
				type={INPUT_TYPES.TEXT}
				onChange={onChange}
				isValid={false}
				value=""
				id="test-input"
			/>
		);
		tree.find('#test-input').simulate('change', {
			target: {
				value: '123',
				type: INPUT_TYPES.TEXT
			}
		});
		expect(onChange).toBeCalledWith({
			error: 'This field is invalid',
			isValid: false,
			value: '123'
		});
	});
});
