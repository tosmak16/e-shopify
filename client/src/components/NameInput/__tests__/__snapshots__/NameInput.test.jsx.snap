// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NameInput renders 1`] = `
<div>
  <TextInput
    className=""
    dark={false}
    disabled={false}
    errorMessage=""
    id="test-input"
    inputDidChange={false}
    isValid={true}
    label=""
    name=""
    onBlur={null}
    onChange={[Function]}
    onChangePasswordVisibilityStatus={[Function]}
    onKeyDown={[Function]}
    placeholder=""
    shouldShowVisibility={false}
    type="text"
    value=""
  />
</div>
`;
